src/psyclone/psyir/transformations/debug_checksum_trans.py

```diff
@@ -0,0 +1,227 @@
+    ...     integer :: i
+    ...     integer :: j
+    ...     do i = 1, 10
+    ...       do j = 1, 10
+    ...         A(i,j) = A(i,k) + i-j
+    ...       end do
+    ...     end do
+    ...     end subroutine
+    ...     """)
+    ... loop = psyir.children[0].children[0]
+    ... DebugChecksumTrans().apply(loop)
+    ... print(FortranWriter()(psyir))
+    subroutine mysubroutine()
+      integer, dimension(10,10) :: a
+      integer :: i
+      integer :: j
+      integer :: PSYCLONE_INTERNAL_line_
+    <BLANKLINE>
+      do i = 1, 10, 1
+        do j = 1, 10, 1
+          a(i,j) = a(i,j) + i - j
+        enddo
+      enddo
+      PSYCLONE_INTERNAL_line_ = __LINE__
+      PRINT *, "PSyclone checksums from mysubroutine at line:", \
+PSYCLONE_INTERNAL_line_ + 1
+      PRINT *, "a checksum", SUM(a)
+    <BLANKLINE>
+    end subroutine mysubroutine
+    <BLANKLINE>
+
+    '''
+    def apply(self, node: Union[Node, List[Node]], options=None) -> None:
+        '''
+        Applies the checksum transformation to the provided node(s).
+
+        :param node: The node or list of nodes to apply the
+                      transformation to.
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node, options={"node-type-check": False})
+
+        node_list = self.get_node_list(node)
+        routine = node_list[0].ancestor(Routine)
+        routine_table = routine.symbol_table
+
+        fwriter = FortranWriter()
+        writes = []
+        # Loop over the assignments in the region
+        assigns = []
+        for node in node_list:
+            assigns.extend(node.walk(Assignment))
+        # Loop through the assignments and find the arrays
+        for assign in assigns:
+            # If we find a structure, we need to check that the final member
+            # is the only array access and is a supported type.
+            if isinstance(assign.lhs, StructureAccessorMixin):
+                # If the reference at assign.lhs is both a Structure
+                # and an Array (e.g. struct(i)%member...) then we
+                # can't know which arrays or indexes to sum over, so
+                # we need to prevent generating a Checksum for this
+                # element.
+                multiple_arrays = isinstance(assign.lhs, ArrayMixin)
+                # Find the last member.
+                member = assign.lhs.member
+                while isinstance(member, StructureAccessorMixin):
+                    # If we have an ArrayMixin Member that isn't the
+                    # final Member in the Structure then we can't
+                    # generate a checksum sensibly, as PSyclone can't know
+                    # which arrays or indexes to sum over.
+                    if (isinstance(member, ArrayMixin) and
+                            member.member is not None):
+                        multiple_arrays = True
```

Comment:
Codecov complains about L146. You could possibly simplify this by using `get_signature_and_indices()` and examining the list of indices - every entry except the last should be empty if we are to support creating a checksum.

Reply:
OK. That's not much (any?) simpler and you've got coverage so I'm happy :-)